game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1221 @@
+
+        this.winProbability = siblingResults.winProbability / gap;
+        this.loseProbability = siblingResults.loseProbability / gap;
+        this.tieProbability = siblingResults.tieProbability / gap;
+        this.averageRounds = siblingResults.averageRounds / gap;
+        this.averageUnits.updateWithChildUnitChances(siblingResults.units, 1.0 / gap);
+
+        // overwrite the cached one with this that has a result
+        nodeCache.put(units, this);
+        // System.out.println("Recurse: " + this);
+      }
+      this.hasResult = true;
+      return true;
+    }
+
+    if (round >= MAX_ROUNDS) {
+      // some branches go for ever
+      // an example is where no one hits each other
+      // treat any of these branches as a tie
+      System.out.println("Reached max rounds");
+      hasResult = true;
+      badProbability = 1.0;
+      // System.out.println("Max: " + this);
+      return true;
+    }
+
+    nodeCache.put(units, this);
+    return false;
+  }
+
+  private void calculateSiblingResultsForRecursiveEnd(
+      final List<BattleStep> steps, final BattleStepSiblings siblingResults) {
+    for (final BattleStep level : steps) {
+      if (level.hasResult) {
+        continue;
+      }
+      siblingResults.probability *= level.probability;
+      siblingResults.winProbability += siblingResults.probability * level.winProbability;
+      siblingResults.loseProbability += siblingResults.probability * level.loseProbability;
+      siblingResults.tieProbability += siblingResults.probability * level.tieProbability;
+      siblingResults.averageRounds += siblingResults.probability * level.averageRounds;
+      if (level.units.getType() != Type.AA_ATTACKER) {
+        // Since the friendly units list never changes, doing an instance equal check will tell
+        // if the units are on the same side
+        if (siblingResults.units.getFriendlyUnits() == level.averageUnits.getFriendlyUnits()) {
+          siblingResults.units.updateFriendlyChances(
+              level.averageUnits.getFriendlyUnitsChances(), siblingResults.probability);
+          siblingResults.units.updateEnemyChances(
+              level.averageUnits.getEnemyUnitsChances(), siblingResults.probability);
+        } else {
+          siblingResults.units.updateFriendlyChances(
+              level.averageUnits.getEnemyUnitsChances(), siblingResults.probability);
+          siblingResults.units.updateEnemyChances(
+              level.averageUnits.getFriendlyUnitsChances(), siblingResults.probability);
+        }
+        calculateSiblingResultsForRecursiveEnd(level.children, siblingResults);
+      }
+    }
+  }
+
+  /**
+   * Figures out all the probabilities of rolling dice with a specific set of powers and amounts.
+   *
+   * @param rollData The details about the dice available to roll
+   * @return List of possible probabilities from rolling 0 dice to rolling rollData.totalRolls})
+   */
+  List<Double> calculateHitProbabilities(final RollData rollData) {
```

Comment:
I think this should be a helper class to simplify the state.

Basically, you would create the object, which will do the initialization in the ctor (all the code of the function) and then call calculateHitProbabilities() on it, which would recurse as necessary. Then, most of the params of calculateHitProbabilitiesWorker() could be removed and could instead be variables on the object.

It will also allow you to make BattleStep shorter by removing this complexity from it.